docs/architecture/adr-004-genesis-modes.md

```diff
@@ -0,0 +1,253 @@
+```
+
+Additionally, a new option function needs to be passed as parameter to `BaseApp` during its initialization:
+
+```go
+// ./baseapp/options.go
+//
+// SetPopulateGenesis sets an option to populate or not the missing values
+// of the each module GenesisState from the mandatory fields given on the genesis file
+func SetPopulateGenesis(pupulateGenesis bool) func(*BaseApp) {
+  return func(bap *BaseApp) { bap.setPopulateGenesis(pupulateGenesis)}
+}
+```
+
+Which requires the following additions to `BaseApp`:
+
+```go
+// ./baseapp/baseapp.go
+type BaseApp struct {
+  // ...
+  populateGenesis bool // flag for pupulating the genesis state from the mandatory fields on genesis.json
+  // ...
+}
+
+func (app *BaseApp) setPopulateGenesis(pupulateGenesis bool) {
+  app.populateGenesis = pupulateGenesis
+}
+```
+
+Finally, the `pupulate-genesis` command imports the provided genesis dile and passes
+the parameter to the module `Manager` call to update the state and then export it:
+
+```go
+// TODO: params
+func PopulateCmd(ctx *Context, appCreator server.AppCreator, appExporter server.AppExporter) *cobra.Command {
+    cmd := &cobra.Command{
+    Use:   "pupulate-genesis [path/to/genesis.json]",
+    Short: "Fill the missing non-mandatory values from the genesis file",
+    Long:  `Fill the missing non-mandatory values from the genesis file by
+    iterating the state`,
+    Args:  cobra.ExactArgs(1),
+    RunE: func(_ *cobra.Command, args []string) error {
+      // initialize args
+      genFile = args[0]
+      genFile := config.GenesisFile()
+
+      if appCreator == nil {
+        return errors.New("app creator constructor function cannot be undefined")
+      }
+
+      if appExporter == nil {
+        return errors.New("app exporter function cannot be undefined")
+      }
+
+      // 1. Import genesis file with missing fields
+      appState, err := codec.MarshalJSONIndent(cdc, mbm.DefaultGenesis())
+      if err != nil {
+        return errors.Wrap(err, "Failed to marshall default genesis state")
+      }
+
+      genDoc := &types.GenesisDoc{}
+      if _, err := os.Stat(genFile); err != nil {
+        if !os.IsNotExist(err) {
+          return err
+        }
+      } else {
+        genDoc, err = types.GenesisDocFromFile(genFile)
+        if err != nil {
+          return errors.Wrap(err, "Failed to read genesis doc from file")
+        }
+      }
+
+      // 2. Create the app and initialize genesis with flag to fill missing values
+      cfg := ctx.Config
+      home := cfg.RootDir
+      traceWriterFile := viper.GetString(flagTraceStore)
+
+      db, err := openDB(home)
+      if err != nil {
+        return nil, err
+      }
+
+      traceWriter, err := openTraceWriter(traceWriterFile)
+      if err != nil {
+        return nil, err
+      }
+
+      // CONTRACT: baseapp must have the option to populate the genesis
+      app := appCreator(ctx.Logger, db, traceWriter)
```

Comment:
I'm confused by this function or why we need to populate the app. This is module manager functionality is it not? We really only need to the BasicModuleManager provided by the app

Reply:
we need to call `app.mm.InitGenesis(ctx, genesisState)` which is called when the app is created through an `AppCreator`, which is application-agnostic (eg not necessarily `SimApp` or `GaiaApp`).

Reply:
ok